### ISSUE TITLE
Is this right?

### ISSUE THREAD

**Issue by blacksythe** [ThePaul-T/WYML#4]
I feel like 8.75% of the thread is more than it should be using

Spark
https://i.imgur.com/UoV9T4d.png

Default configuration from FTB Oceanblock

wyml.json
{
	// Set the max amount of spawn tries base rate
	"MOB_TRIES": 4,
	// Set to true to multiply base rate by online player count
	"MULTIPLY_BY_PLAYERS": true,
	// Replace Mojang magic number (pow2) with this
	"MOJANG_MAGIC_NUM": 17.0,
	// Spam your console and make performance terrible...
	"DEBUG_PRINT": false,
	// Downscale the Mojang Magic Number by the online player count
	"DOWNSCALE_MAGIC_NUM": true,
	// If downscaling enabled, do not allow below this value
	"DOWNSCALE_MAGIC_NUM_MIN": 10.0,
	// Amount of nano seconds to wait for a task (Mojang default is 100000), thread locking uses the futex[https://man7.org/linux/man-pages/man2/futex.2.html] syscall on linux, causing a lot of syscalls when the value is too low
	"TASK_WAIT_NANOS": 5000000,
	// Enable Minecraft dataFixerUpper (enables you to upgrade worlds between Minecraft versions), disabling this stops the big ram spike at server start when loading existing worlds [ENABLE THIS IF YOU HAVE ANOTHER DFU CHANGING MOD!]
	"ENABLE_DFU": true,
	// Force Java garbage collector to run once the levels have been generated or loaded the first time (Frees up memory after the server is initially started)
	"ENABLE_GARBAGE_COLLECTION_LOAD": true,
	// Ensure the tick loop does not run repeatedly, waits until the next tick is due (reduce cpu usage on hardware)
	"NORMALIZE_TICKS": true,
	// Allow pausing of spawns in specific chunks
	"ALLOW_PAUSE": true,
	// Allow pausing of chunks claimed using FTB Chunks
	"ALLOW_PAUSE_CLAIMED": false,
	// Allow pausing of force chunk loaded chunks
	"ALLOW_PAUSE_FORCED": false,
	// At what what percentage of failed spawns should we then pause spawning in a chunk
	"PAUSE_RATE": 65,
	// How long to pause spawning if pause spawn failure rate reached
	"PAUSE_TICKS": 1800,
	// At what what percentage of successful spawns, after we pause spawning, should we resume spawning
	"RESUME_RATE": 10,
	// At what what percentage of failed spawns should we then pause spawning in a chunk in claimed chunks
	"PAUSE_CLAIMED_RATE": 80,
	// How long to pause spawning if pause spawn failure rate reached in claimed chunks
	"PAUSE_CLAIMED_TICKS": 150,
	// At what what percentage of successful spawns, after we pause spawning, should we resume spawning in claimed chunks
	"RESUME_CLAIMED_RATE": 4,
	// The minimum amount of attempted spawns of a type in a chunk before we allow pausing
	"PAUSE_MIN": 256,
	// Allow slowing of spawns in specific chunks
	"ALLOW_SLOW": true,
	// Maximum spawn requests per chunk per tick based off average spawn rate in sample spawn rate
	"MAX_CHUNK_SPAWN_REQ_TICK": 12,
	// How long to stay in slow mode after spawn rates are under control
	"SLOW_TICKS": 600,
	// How many ticks to sample (and average) spawn rates over
	"SAMPLE_TICKS": 5,
	// How many ticks to remember if we fail spawning in a block position
	"SPAWNLOC_CACHE_TICKS": 400,
	// How many ticks to store a SpawnManager for a chunk after it's last update
	"MANAGER_CACHE_TICKS": 600,
	// Disable this once you are happy with the configs you have and their impact on memory. This will print how many SpawnManagers and SpawnCaches we have at once, every 10 seconds.
	"CLEAN_PRINT": false
}

**Comment by ThePaul-T** [comment]
Probably is right, considering the majority of the CPU users there are from copies (with modifications) of the main Mojang chunk tick functions, so a lot of the chunk work is going on there.

This was designed around a real-world issue with multiplayer on OceanBlock, so, if you're playing single-player, there'll be no ill effects to disabling the mod honestly, but there should also be no real perceivable issue with leaving it there.

Basically, some map types try and spawn mobs to mob-cap repeatedly every tick due to there being no actual land to spawn on, so after a few players it reaches a point the server crawls to a standstill, this tracks that behaviour and disables spawns in chunks where it can't find anywhere to spawn mobs.

Not really any other sane way to handle it without disabling spawns entirely and using spawners, or just rewriting the whole of Mojangs spawning logic.

**Comment by blacksythe** [comment]
Thanks, Will do some testing with another mod that also does a similar thing, but as this is on a server, you can imagine trying to keep a steady 17+ tps is an ongoing battle

**Comment by ThePaul-T** [comment]
I can, basically, OceanBlock used to hit sub 10tps with 6 people on before I wrote this, this was just everything I could think to do to make things at least usable, I'll progress with it and change/add more configs as time goes on, but there is a trade-off between ram and CPU to do things and it's a hard balance.

Honestly, I'd say just adjust some configs, I tried to make every single option changeable so people can adjust this to what is needed.
"ALLOW_PAUSE_CLAIMED": false, to true would likely fix your tps, but no mobs are probably going to spawn unless mob farms are a full chunk in size.

"ALLOW_PAUSE_FORCED": false, should probably be true anyway on a server.
